### DISCUSSION TITLE
Returning current json object when using `json[""]` or `json['.']`

### DISCUSSION THREAD

**Discussion by LecrisUT** [nlohmann/json#3113] [Q&A] [answered]
I'm sorry if this is already documented or discussed, but in the current implementation, does using `json[""]`, i.e. with empty string or something generally invalid like `json['.']` return the current json object?

I want to reuse part of my serializer, where I have something as follows:
```cpp
class A {
    json& j;
    json& GetRef(const std::string& name, int ind = -1){
        return ind<0 ? j[name] : j[name][ind];
    }
}
```
I'm trying to simplify `GetRef` so it doesn't need to have another check against `name` to access `j[ind]` when the reference `j` is at an array item.

**Comment by nlohmann** [comment, marked as answer]
I do not understand what you're trying to achieve.

Assuming `j` is a JSON object, then `j[""]` returns a reference to the value stored at key `""` (which is valid in JSON), or, if this key does not exist, creates a `null` value for this key and returns a reference to this value. See https://json.nlohmann.me/api/basic_json/operator%5B%5D/ for more information.

**Reply by LecrisUT** [reply]
Thank you for the confirmation, then I would have to add a check on the implementation side, to achieve what I want. For reference, consider the following json:
```json
{
  "A" : {
            "B" : { ... },
            "C" : [ { ... }, {...}]
        }
}
```
Let's say that currently I have a reference `j=baseJson["A"]`. Accessing `B` and  array items of `C` is no problem with the above `GetRef` as they would call `j["B"]` and `j["C"][ind]`. But how about if `j=baseJson["A"]["C"]`? Then I would have to use `j[ind]`. It would be more convenient to be able to pass something like `./` for directory locations, but for json instead.

A more flushed out example of how I use this is like this:
```cpp
virtual void LoadData( A& Tracker, const std::string& name, std::shared_ptr<obj>& data){
   // Much preparations
   data.reset(new obj(Tracker[name][var1].get<...>(), ... ));
}
virtual void LoadData( A& Tracker, const std::string& name, std::vector<std::shared_ptr<obj>>& data){
   for ( auto i = 0; i < Tracker[name].size(); i++ ){
      shared_ptr p;
      auto Tracker2 = Tracker;
      Tracker2.j = Tracker.j[name][i];
      LoadData( Tracker2, "", p );
      data.push_back(p);
   }
}
```
This I would use to load both `B` object and the array of `C`. I would prefer to limit copying the code for `int ind` instead of `std::string name` because of the preparation part, and I would prefer not to encapsulate everything in a templates for that.

PS: there are other sections similar sections like that with different preparation steps as well.

**Reply by nlohmann** [reply]
I still have not fully understood the issue. But 

> able to pass something like ./ for directory locations, but for json instead.

reminds me of JSON Pointer which are supported by the library, see https://json.nlohmann.me/features/json_pointer/.

**Reply by LecrisUT** [reply]
Indeed that's a good alternative to design around. Maybe if others find uses for this proposal it can be revisited.
